code/modules/projectiles/guns/energy/modular.dm

```diff
@@ -0,0 +1,282 @@
+			item_state = "retro"
+		if(CHASSIS_MEDIUM)
+			gun_type = CHASSIS_MEDIUM
+			slot_flags = SLOT_BELT | SLOT_BACK
+			item_state = "energystun"
+		if(CHASSIS_LARGE)
+			gun_type = CHASSIS_LARGE
+			slot_flags = SLOT_BACK
+			item_state = "lasercannon"
+
+	if((capacitor.reliability - capacitor.condition <= 0))
+		qdel(capacitor)
+		capacitor = null
+		world << "wat"
+
+	if(focusing_lens.reliability - focusing_lens.condition <= 0)
+		qdel(focusing_lens)
+		focusing_lens = null
+		world << "wat2"
+
+	if(!focusing_lens || !capacitor || !modulator)
+		disassemble()
+		return
+
+	reliability = (capacitor.reliability - capacitor.condition) + (focusing_lens.reliability - focusing_lens.condition)
+
+	if(modulator.projectile)
+		projectile_type = modulator.projectile
+
+	fire_delay = capacitor.fire_delay
+	max_shots = capacitor.shots
+	power_supply.maxcharge = max_shots*charge_cost
+	dispersion = focusing_lens.dispersion
+	accuracy = focusing_lens.accuracy
+	burst += focusing_lens.burst
+
+	if(gun_mods.len)
+		handle_mod()
+
+	fire_delay_wielded = min(0,(fire_delay - fire_delay*3))
+	accuracy_wielded = accuracy + accuracy/4
+	scoped_accuracy = accuracy_wielded + accuracy/4
+	max_shots = max_shots * burst
+	w_class = gun_type
+	if(reliability <= 0)
+		reliability = 1
+
+/obj/item/weapon/gun/energy/laser/prototype/proc/handle_mod()
+	for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+		switch(modifier.mod_type)
+			if(MOD_SILENCE)
+				silenced = 1
+			if(MOD_NUCLEAR_CHARGE)
+				self_recharge = 1
+				criticality *= 2
+		fire_delay *= modifier.fire_delay
+		reliability += modifier.reliability
+		burst += modifier.burst
+		burst_delay += modifier.burst_delay
+		max_shots *= modifier.shots
+		force += modifier.gun_force
+		chargetime += modifier.chargetime*10
+		accuracy += modifier.accuracy
+		criticality *= modifier.criticality
+		if(modifier.scope_name)
+			zoomdevicename = modifier.scope_name
+
+/obj/item/weapon/gun/energy/laser/prototype/consume_next_projectile()
+	if(!power_supply)
+		return null
+	if(!ispath(projectile_type))
+		return null
+	if(!power_supply.checked_use(charge_cost))
+		return null
+	if (self_recharge)
+		addtimer(CALLBACK(src, .proc/try_recharge), recharge_time * 2 SECONDS, TIMER_UNIQUE)
+	var/obj/item/projectile/beam/A = new projectile_type(src)
+	A.damage = capacitor.damage
+	for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+		A.damage *= modifier.damage
+	A.damage = A.damage/(burst - 1)
+	A.damage *= modulator.damage
+	for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+		if(prob(modifiers.len * 20))
+			capacitor.degrade(modifier.malus)
+		if(prob(33))
+			modifier.degrade(1)
+	focusing_lens.degrade(capacitor.damage += modulator.damage)
+
+	updatetype()
+	return A
+
+/obj/item/weapon/gun/energy/laser/prototype/proc/disassemble()
+	if(gun_mods.len)
+		for(var/obj/item/laser_components/modifier/modifier in gun_mods)
+			modifier.forceMove(get_turf(src))
```

Comment:
At the top of this method:
```dm
var/turf/T = get_turf(src)
```

And then just do `x.forceMove(T)`. Calling get_turf this often is silly.